### ISSUE TITLE
Manage Kinds using Terraform

### ISSUE THREAD

**Issue by atalema** [launchdarkly/terraform-provider-launchdarkly#183]
Is there a way to manage to LaunchDarkly context kinds using the terraform LD provider?

**Comment by ldhenry** [comment]
Hey @atalema,

Currently this is not possible today, but we do have plans to add support for this in the future. We'll be sure to update this issue when it's available.

Thanks,
Henry

**Comment by atalema** [comment]
Thanks Henry for getting back to me. Do you've a suggestion on how to manage kinds using terraform, till we get the support using LD provider? One idea came to my mind, is to use terraform HTTP provider to communicate with LD API https://registry.terraform.io/providers/hashicorp/http/latest/docs/data-sources/http any thoughts on this one?

**Comment by ldhenry** [comment]
I don't have an experience with the HTTP provider but it looks like you can only use it to make GET requests. This could be useful if you want to reference an existing context kind in your Terraform config, however you would need to use something else to create and update the context kinds. 

Do you have an idea on the number of context kinds you plan to manage? If the number is small you could use the LaunchDarkly UI to create and update your context kinds until the new Terraform provider resource is available.

**Comment by atalema** [comment]
you're right, it's allows POST but for readonly urls. Which means only supports GET scenarios. Alternatively, I found this provider https://registry.terraform.io/providers/Mastercard/restapi/latest/docs which is a full rest api provider. I believe I can use to create / update the context kind. Another option, I could write a python script which could do the job creating the kind. 

They're not huge it could end up as 20+ but the problem is not about the number, is around:

- How can we maintain some consistency across, so we don't end up with different kinds trying to achieve the same thing. Which it would end up eating our MAU quota quickly.
- How can we govern these changes through Infra-structure as code. We do today allow creating projects only through raising a PR for our terraform project. Which enables us to have a point of discussion with teams. 

Thanks for taking the time to give your thoughts on this issue.

**Comment by ldhenry** [comment]
That makes sense. For now, I would recommend enforcing a custom role on most members that prevents them from creating context kinds. Hopefully this can tide you over until we have a dedicated terraform provider resource available.

**Comment by EDN-liamkelly** [comment]
Any progress on this?

**Comment by monsagri** [comment]
Hi @EDN-liamkelly and @atalema 

I realise it's been a long time coming, but w're preparing to release v3 of the Provider, which will implement this among a number of other features.

The shape of context kind is as below, they are keyed by context kind key, and we've got a data source as well.
```
resource "launchdarkly_context_kind" "organization" {
  project_key = launchdarkly_project.example.key
  key         = "organization"
  name        = "Organization"
  description = "An organization that owns one or more accounts"
}
```

As it is a beta release, you will need to explicitly select the preview version. There are some migration steps involved due to breaking changes, however these are documented [here](https://registry.terraform.io/providers/launchdarkly/launchdarkly/latest/docs/guides/migrating-to-v3) and the provider ships with migrate-tf-syntax, which automates the majority of the HCL migration. There's also a skill if you'd like to use an [agent to assist](https://github.com/launchdarkly/terraform-provider-launchdarkly/blob/preview-v3/.claude/skills/terraform-provider-v3-migration/SKILL.md).

If you're interested in trying it out, I'd love to hear your thoughts. Either way I'll come back to this issue once we release v3.

Thanks!

**Comment by monsagri** [comment]
This is resolved in the V3 Release of the provider.   

[Github Release
](https://github.com/launchdarkly/terraform-provider-launchdarkly/releases/tag/v3.0.0)
[Terraform Registry Docs](https://registry.terraform.io/providers/launchdarkly/launchdarkly/latest/docs/guides/migrating-to-v3)
